src/mlpack/core/math/trigamma.hpp

```diff
@@ -0,0 +1,211 @@
+                     -3.3703848401898283,
+                     0.28080574467981213,
+                     1.6638069578676164,
+                     0.64468386819102836
+                    };
+
+  const T Q_1_2[] = {
+                     1.0,
+                     3.4535389668541151,
+                     4.5208926987851437,
+                     2.7012734178351534,
+                     0.64468798399785611,
+                     -0.20314516859987728e-6
+                    };
+
+  const T P_2_4[] = {
+                     -0.13803835004508849e-7,
+                     0.50000049158540261,
+                     1.6077979838469348,
+                     2.5645435828098254,
+                     2.0534873203680393,
+                     0.74566981111565923
+                    };
+
+  const T Q_2_4[] = {
+                     1.0,
+                     2.8822787662376169,
+                     4.1681660554090917,
+                     2.7853527819234466,
+                     0.74967671848044792,
+                     -0.00057069112416246805
+                    };
+
+  const T P_4_INF[] = {
+                       0.68947581948701249e-17L,
+                       0.49999999999998975L,
+                       1.0177274392923795L,
+                       2.498208511343429L,
+                       2.1921221359427595L,
+                       1.5897035272532764L,
+                       0.40154388356961734L
+                      };
+
+  const T Q_4_INF[] = {
+                       1.0L,
+                       1.7021215452463932L,
+                       4.4290431747556469L,
+                       2.9745631894384922L,
+                       2.3013614809773616L,
+                       0.28360399799075752L,
+                       0.022892987908906897L
+                      };
+
+  // For 1 < x <= 2
+  if (x <= 2)
+    return (offset + EvaluatePolyPrec(P_1_2, x) / EvaluatePolyPrec(Q_1_2, x)) / (x * x);
+  // For 2 < x <= 4
+  else if (x <= 4)
+  {
+    T y = 1 / x;
+    return (1 + EvaluatePolyPrec(P_2_4, y) / EvaluatePolyPrec(Q_2_4, y)) / x;
+  }
+
+  // For x > 4
+  T y = 1 / x;
+  return (1 + EvaluatePolyPrec(P_4_INF, y) / EvaluatePolyPrec(Q_4_INF, y)) / x;
+}
+
+/**
+ * This function calculates and returns Trigamma(x).
+ * For x < 0 we use reflection.
+ * For 0 < x < 1 bring x to interval [1, INF].
+ * 
+ * @param x Input for which we have to calculate trigamma.
+ */
+template<typename T>
+T Trigamma(T x)
+{
+  T result = 0;
+
+  if (x <= 0)
+  {
+    // Reflect
+    T z = 1 - x;
+
+    // Check for evaluation on pole
+    if (floor(x) == x)
+      throw std::runtime_error("Evaluation of the function at pole.");
+    
+    T s = fabs(x) < fabs(z) ? sin(M_PI * x) : sin(M_PI * z);
+    
+    return -Trigamma(z) + pow(M_PI, 2) / (s * s);
+  }
+
+  if (x < 1)
+  {
+    // Make x > 1
+    result = 1 / (x * x);
+    x += 1;
+  }
+
+  // For x >= 1
+  return result + TrigammaPrec(x);
+}
+
+} // namespace math
+} // namespace mlpack
+
+#endif
+
```

Comment:
```suggestion
#endif
```
No need for the empty line. :)

Reply:
```suggestion
#endif
```